asapdiscovery-data/asapdiscovery/data/state_expanders/protomer_expander.py

```diff
@@ -0,0 +1,213 @@
+import os
+import subprocess
+import tempfile
+from typing import Literal
+
+from asapdiscovery.data.openeye import (
+    load_openeye_sdfs,
+    oechem,
+    oequacpac,
+    save_openeye_sdfs,
+)
+from asapdiscovery.data.schema_v2.ligand import Ligand
+from asapdiscovery.data.state_expanders.state_expander import StateExpanderBase
+from pydantic import Field
+
+
+class ProtomerExpander(StateExpanderBase):
+    """
+    Expand a molecule to protomers using OpenEye reasonable protomer state enumeration.
+
+    Note:
+        The input molecule is included in the output.
+    """
+
+    expander_type: Literal["ProtomerExpander"] = "ProtomerExpander"
+
+    def _provenance(self) -> dict[str, str]:
+        return {
+            "oechem": oechem.OEChemGetVersion(),
+            "quacpac": oequacpac.OEQuacPacGetVersion(),
+        }
+
+    def _expand(self, ligands: list[Ligand]) -> list[Ligand]:
+        expanded_states = []
+        provenance = self.provenance()
+        for parent_ligand in ligands:
+            oemol = parent_ligand.to_oemol()
+            for protomer in oequacpac.OEGetReasonableProtomers(oemol):
+                fmol = oechem.OEMol(protomer)
+                # copy the ligand properties over to the new molecule, we may want to have more fine grained control over this
+                # down the track.
+                protomer_ligand = Ligand.from_oemol(fmol, **parent_ligand.dict())
+                if protomer_ligand.fixed_inchikey != parent_ligand.fixed_inchikey:
+                    # only add tags to new microstates of the input molecule
+                    protomer_ligand.set_expansion(
+                        parent=parent_ligand, provenance=provenance
+                    )
+                    expanded_states.append(protomer_ligand)
+                else:
+                    expanded_states.append(parent_ligand)
+            # add the parent if it is not present.
+            if parent_ligand not in expanded_states:
+                expanded_states.append(parent_ligand)
+
+        return expanded_states
+
+
+class EpikExpander(StateExpanderBase):
+    """
+    Expand the protomer and tautomeric states of a molecule using epik and capture the state penalties.
+
+    Note:
+        The method assumes you have schrodinger software installed and the path to the software is exported as a
+        environment variable named SCHRODINGER.
+    """
+
+    expander_type: Literal["EpikExpander"] = "EpikExpander"
+
+    ph: float = Field(
+        7.3,
```

Comment:
Just me being picky. I saw in other places mentions of ph 7.4, should we be consistent and use 7.4 here as well? Not a blocker.

Reply:
This is taken from the old scripts to make sure we get the same expected outputs. So I think we should keep it as this for now but it is configurable in the workflow so easy to change during a run. 
https://github.com/choderalab/perses/blob/6c6419f96878be3003673f68d2e34d70826191dc/examples/moonshot-mainseries/02-generate-poses.py#L197